### PR TITLE
Add support for Alipan Storage backend

### PR DESCRIPTION
**Pull request by PeterDing**
Add support to access [Alipan Storage](https://www.alipan.com/).

## Description of changes:

* Add `download_from_alipan` function and `AlipanUploader` class

* Add tests for both download and upload

* Add docs for configure Alipan Storage Credentials

* Add `AliPCS-Py` to setup.py as extra dependence

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -119,6 +119,10 @@
     'databricks-sdk==0.23.0',
 ]
 
+extra_deps['alipan'] = [
+    'AliPCS-Py>=0.8',
```
Comment (karan6181) [inline setup.py:123]: Can you please restrict the version to a major version? For example, `AliPCS-Py>=0.8,<1`?

File: streaming/base/storage/download.py [hunk 2]
```diff
@@ -422,6 +423,47 @@ def download_from_dbfs(remote: str, local: str) -> None:
...
+
+    api = AliPCSApiMix(web_refresh_token, web_token_type=web_token_type)
+
+    obj = urllib.parse.urlparse(remote)
+    if obj.scheme != 'alipan':
+        raise ValueError(
+            f'Expected obj.scheme to be `alipan`, instead, got {obj.scheme} for remote={remote}')
+    if obj.netloc != '':
+        raise ValueError(
+            f'Expected remote is like alipan:///path/to/some, instead, got remote={remote}')
```
Comment (karan6181) [inline streaming/base/storage/download.py:448]:
```suggestion
            f'Expected remote to be alipan://path/to/some, instead, got remote={remote}')
```

File: streaming/base/storage/download.py [hunk 3]
```diff
@@ -422,6 +423,47 @@ def download_from_dbfs(remote: str, local: str) -> None:
...
+
+    remote_path = pathlib.Path(obj.path)
+    filename = remote_path.name
+    localdir = pathlib.Path(local).parent
+
+    try:
```
Comment (karan6181) [inline streaming/base/storage/download.py:454]: Wondering, what's the point of `try..except` if you are just raising the same exception in the `except` block?

File: streaming/base/storage/upload.py [hunk 4]
```diff
@@ -945,6 +947,137 @@ def check_folder_exists(self):
...
+        super().__init__(out, keep_local, progress_bar, retry, exist_ok)
+
+        obj = urllib.parse.urlparse(self.remote)
+        if obj.scheme != 'alipan':
+            raise ValueError(
+                f'Expected obj.scheme to be `alipan`, instead, got {obj.scheme} for remote={self.remote}'
+            )
+        if obj.netloc != '':
+            raise ValueError(
+                f'Expected remote is like alipan:///path/to/some, instead, got remote={self.remote}'
```
Comment (karan6181) [inline streaming/base/storage/upload.py:986]:
```suggestion
                f'Expected remote to be alipan://path/to/some, instead, got remote={self.remote}'
```

File: streaming/base/storage/upload.py [hunk 5]
```diff
@@ -945,6 +947,137 @@ def check_folder_exists(self):
...
+
+        web_refresh_token = os.environ['ALIPAN_WEB_REFRESH_TOKEN']
+        web_token_type = 'Bearer'
+        self.alipan_encrypt_password = os.environ.get('ALIPAN_ENCRYPT_PASSWORD', '').encode()
+        self.alipan_encrypt_type = EncryptType.No  # No encryption by default
+        encrypt_type = os.environ.get('ALIPAN_ENCRYPT_TYPE', '')
+        if encrypt_type:
+            if getattr(EncryptType, encrypt_type, None) is None:
+                raise ValueError(
+                    f'Invalid ALIPAN_ENCRYPT_TYPE: {encrypt_type}. Encryption type must be one of `Simple`, `ChaCha20`, `AES256CBC`'
```
Comment (karan6181) [inline streaming/base/storage/upload.py:1000]: Please split this line. It's greater than 99 characters.

File: streaming/base/storage/upload.py [hunk 6]
```diff
@@ -945,6 +947,137 @@ def check_folder_exists(self):
...
+        Args:
+            prefix (Optional[str], optional): The prefix to search for. Defaults to ``None``.
+
+        Returns:
+            List[str]: A list of object names that match the prefix.
+        """
+        if prefix is None:
+            prefix = ''
+
+        assert self.remote is not None
```
Comment (karan6181) [inline streaming/base/storage/upload.py:1065]: Please avoid `assert`, instead raise a python exception with a descriptive error message.

**Review by karan6181 — COMMENTED** [review]
Thank you for your contribution. Posted some minor comments.

File: streaming/base/storage/download.py (hunk 3 above)
Comment (PeterDing) [reply streaming/base/storage/download.py:454]: A `FileNotFoundError` should be raised here.

File: streaming/base/storage/upload.py (hunk 4 above)
Comment (PeterDing) [reply streaming/base/storage/upload.py:986]: The Alipan url doesn't need a `netloc` part. The `path` part is only needed. So I give the url as `alipan:///path/to/some`. Is it ok?

File: streaming/base/storage/upload.py (hunk 5 above)
Comment (PeterDing) [reply streaming/base/storage/upload.py]: done.

File: streaming/base/storage/upload.py (hunk 6 above)
Comment (PeterDing) [reply streaming/base/storage/upload.py]: I give a `ValueError` when `self.remote` is None.

File: setup.py (hunk 1 above)
Comment (PeterDing) [reply setup.py]: done.

File: streaming/base/storage/download.py (hunk 3 above)
Comment (PeterDing) [reply streaming/base/storage/download.py]: done.

File: streaming/base/storage/download.py [hunk 7]
```diff
@@ -474,6 +519,8 @@ def download_file(remote: Optional[str], local: str, timeout: float):
         download_from_databricks_unity_catalog(remote, local)
     elif remote.startswith('dbfs:/'):
         download_from_dbfs(remote, local)
+    elif remote.startswith('alipan://'):
```
[the remark this answers is not in the document]
Comment (karan6181) [reply streaming/base/storage/download.py:522]: Can you please share one example of a full remote path?

[the remark this answers is not in the document]
Comment (PeterDing) [reply streaming/base/storage/download.py:522]: In Alipan, a full remote path is a normal POSIX system path like `/dir1/dir2/somefile`.
So, `alipan:///dir1/dir2/somefile` is for streaming with the prefix `alipan://`. No need the `netloc` of the url.

[the remark this answers is not in the document]
Comment (snarayan21) [reply streaming/base/storage/download.py:522]: Hey @PeterDing, just getting back to this now. `alipan://` should be fine.

File: docs/source/how_to_guides/configure_cloud_storage_credentials.md [hunk 8]
```diff
@@ -274,3 +274,50 @@ export DATABRICKS_HOST='hostname'
 export DATABRICKS_TOKEN='token key'
 ```
 ````
+
+
+## Alipan
+
+To authenticate Alipan access, users must set their Alipan refresh token (`ALIPAN_WEB_REFRESH_TOKEN`) in the run environment.
+
+Get the refresh token from the Alipan website, user needs to login to the [Alipan website](https://www.alipan.com/drive) and go to the console of browser's devTools and pass the below code in the console to get the refresh token.
```
Comment (snarayan21) [inline docs/source/how_to_guides/configure_cloud_storage_credentials.md:283]:
```suggestion
To get the refresh token from the Alipan website, the user needs to login to the [Alipan website](https://www.alipan.com/drive), go to the console of the browser's devTools, and pass the code below into the console to get the refresh token.
```

File: docs/source/how_to_guides/configure_cloud_storage_credentials.md [hunk 9]
```diff
@@ -274,3 +274,50 @@ export DATABRICKS_HOST='hostname'
...
+```
+
+```{code-tab} sh
+export ALIPAN_WEB_REFRESH_TOKEN='refresh_token'
+```
+````
+
+### Encryption Data
+
+If you want to encrypt the data in the cloud storage, you can set the below environment variable to encrypt and decrypt the data.
```
Comment (snarayan21) [inline docs/source/how_to_guides/configure_cloud_storage_credentials.md:304]:
```suggestion
If you want to encrypt data in cloud storage, you can set the environment variable below to encrypt and decrypt the data.
```

**Review by snarayan21 — COMMENTED** [review]
Hey @PeterDing, just getting back to this now. Left a couple of suggestions and comments. Are there any other changes you'd like to make?

**Review by snarayan21 — APPROVED** [review]
lgtm! this will be included in the upcoming release -- thanks for contributing!

**Comment by snarayan21** [comment]
@PeterDing It would be most helpful if we could refer users to you if they face issues with Alipan. You are much more well versed with Alipan stuff than our team :)

Excited to get this integrated into Streaming!

**Pull request merged**